### ISSUE TITLE
[CupertinoDatePicker] Showing months incorrectly in pt-br

### ISSUE THREAD

**Issue by jehhxuxu** [flutter/flutter#138305]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Just create an app with supportedLocales: const [Locale('pt', 'BR')], and use CupertinoDatePicker

### Expected results

I think that the expected result is the month word shown correctly, maybe we can calculate the size of the word instead of the length of the word.

### Actual results

Currently, CupertinoDatePicker is calculating the space of the month column based on the longestText. In pt-br the longer month is Fevereiro but Dezembro is larger in space because we have larger letters. With that, the CupertinoDatePicker is cutting the month word. Even if we set dateOrder mdy or dmy, the problem persists. 

![Screenshot 2023-11-11 at 16 57 40](https://github.com/flutter/flutter/assets/27310692/e7281688-6aae-4770-8e83-c55cac3d9df6)

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.13.5, on macOS 14.0 23A344 darwin-x64, locale en-BR)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] IntelliJ IDEA Community Edition (version 2023.2.3)
[✓] Connected device (4 available)
[✓] Network resources

• No issues found!

```

</details>

**Comment by danagbemava-nc** [comment]
Reproducible on the latest stable and master. 

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_localizations/flutter_localizations.dart';

/// Flutter code sample for [CupertinoDatePicker].

void main() => runApp(const DatePickerApp());

class DatePickerApp extends StatelessWidget {
  const DatePickerApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const CupertinoApp(
      theme: CupertinoThemeData(brightness: Brightness.light),
      locale: Locale('pt', 'BR'),
      supportedLocales: [
        Locale('pt', 'BR'),
      ],
      localizationsDelegates: [
        GlobalMaterialLocalizations.delegate,
        GlobalWidgetsLocalizations.delegate,
        GlobalCupertinoLocalizations.delegate,
      ],
      home: DatePickerExample(),
    );
  }
}

class DatePickerExample extends StatefulWidget {
  const DatePickerExample({super.key});

  @override
  State<DatePickerExample> createState() => _DatePickerExampleState();
}

class _DatePickerExampleState extends State<DatePickerExample> {
  DateTime date = DateTime(2016, 10, 26);
  DateTime time = DateTime(2016, 5, 10, 22, 35);
  DateTime dateTime = DateTime(2016, 8, 3, 17, 45);

  // This function displays a CupertinoModalPopup with a reasonable fixed height
  // which hosts CupertinoDatePicker.
  void _showDialog(Widget child) {
    showCupertinoModalPopup<void>(
      context: context,
      builder: (BuildContext context) => Container(
        height: 216,
        padding: const EdgeInsets.only(top: 6.0),
        // The Bottom margin is provided to align the popup above the system
        // navigation bar.
        margin: EdgeInsets.only(
          bottom: MediaQuery.of(context).viewInsets.bottom,
        ),
        // Provide a background color for the popup.
        color: CupertinoColors.systemBackground.resolveFrom(context),
        // Use a SafeArea widget to avoid system overlaps.
        child: SafeArea(
          top: false,
          child: child,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      navigationBar: const CupertinoNavigationBar(
        middle: Text('CupertinoDatePicker Sample'),
      ),
      child: DefaultTextStyle(
        style: TextStyle(
          color: CupertinoColors.label.resolveFrom(context),
          fontSize: 22.0,
        ),
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              _DatePickerItem(
                children: <Widget>[
                  const Text('Date'),
                  CupertinoButton(
                    // Display a CupertinoDatePicker in date picker mode.
                    onPressed: () => _showDialog(
                      CupertinoDatePicker(
                        initialDateTime: date,
                        mode: CupertinoDatePickerMode.date,
                        use24hFormat: true,
                        // This shows day of week alongside day of month
                        showDayOfWeek: true,
                        // This is called when the user changes the date.
                        onDateTimeChanged: (DateTime newDate) {
                          setState(() => date = newDate);
                        },
                      ),
                    ),
                    // In this example, the date is formatted manually. You can
                    // use the intl package to format the value based on the
                    // user's locale settings.
                    child: Text(
                      '${date.month}-${date.day}-${date.year}',
                      style: const TextStyle(
                        fontSize: 22.0,
                      ),
                    ),
                  ),
                ],
              ),
              _DatePickerItem(
                children: <Widget>[
                  const Text('DateTime'),
                  CupertinoButton(
                    // Display a CupertinoDatePicker in dateTime picker mode.
                    onPressed: () => _showDialog(
                      CupertinoDatePicker(
                        initialDateTime: dateTime,
                        use24hFormat: true,
                        // This is called when the user changes the dateTime.
                        onDateTimeChanged: (DateTime newDateTime) {
                          setState(() => dateTime = newDateTime);
                        },
                      ),
                    ),
                    // In this example, the time value is formatted manually. You
                    // can use the intl package to format the value based on the
                    // user's locale settings.
                    child: Text(
                      '${dateTime.month}-${dateTime.day}-${dateTime.year} ${dateTime.hour}:${dateTime.minute}',
                      style: const TextStyle(
                        fontSize: 22.0,
                      ),
                    ),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}

// This class simply decorates a row of widgets.
class _DatePickerItem extends StatelessWidget {
  const _DatePickerItem({required this.children});

  final List<Widget> children;

  @override
  Widget build(BuildContext context) {
    return DecoratedBox(
      decoration: const BoxDecoration(
        border: Border(
          top: BorderSide(
            color: CupertinoColors.inactiveGray,
            width: 0.0,
          ),
          bottom: BorderSide(
            color: CupertinoColors.inactiveGray,
            width: 0.0,
          ),
        ),
      ),
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 16.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: children,
        ),
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[!] Flutter (Channel stable, 3.13.9, on macOS 14.0 23A344 darwin-arm64, locale en-GB)
    • Flutter version 3.13.9 on channel stable at /Users/nexus/dev/sdks/flutter
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutters/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter. Consider adding /Users/nexus/dev/sdks/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutters/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter. Consider adding /Users/nexus/dev/sdks/flutter/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d211f42860 (3 weeks ago), 2023-10-25 13:42:25 -0700
    • Engine revision 0545f8705d
    • Dart version 3.1.5
    • DevTools version 2.25.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc1)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0-rc1
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode-15.0.0-Release.Candidate.app/Contents/Developer
    • Build 15A240d
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.3)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 76.0.4
    • Dart plugin version 232.10072.19

[✓] VS Code (version 1.84.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device (4 available)
    • Nexus (mobile)       • 00008020-001875E83A38002E • ios            • iOS 17.1.1 21B91
    • Dean’s iPad (mobile) • 00008103-000825C811E3401E • ios            • iOS 17.1 21B74
    • macOS (desktop)      • macos                     • darwin-arm64   • macOS 14.0 23A344 darwin-arm64
    • Chrome (web)         • chrome                    • web-javascript • Google Chrome 119.0.6045.123

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

```
[✓] Flutter (Channel master, 3.17.0-5.0.pre.64, on macOS 14.0 23A344 darwin-arm64, locale en-GB)
    • Flutter version 3.17.0-5.0.pre.64 on channel master at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2e088178e8 (9 hours ago), 2023-11-12 18:45:26 -0500
    • Engine revision a18ee3c7f5
    • Dart version 3.3.0 (build 3.3.0-120.0.dev)
    • DevTools version 2.30.0-dev.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc1)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0-rc1
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode-15.0.0-Release.Candidate.app/Contents/Developer
    • Build 15A240d
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.3)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 76.0.4
    • Dart plugin version 232.10072.19

[✓] VS Code (version 1.84.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device (4 available)
    • Nexus (mobile)       • 00008020-001875E83A38002E • ios            • iOS 17.1.1 21B91
    • Dean’s iPad (mobile) • 00008103-000825C811E3401E • ios            • iOS 17.1 21B74
    • macOS (desktop)      • macos                     • darwin-arm64   • macOS 14.0 23A344 darwin-arm64
    • Chrome (web)         • chrome                    • web-javascript • Google Chrome 119.0.6045.123

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```
</details>

**Comment by MitchellGoodwin** [comment]
I looked into this a bit, putting my notes here. As mentioned, the issue is that for every column, the longest word is being found by amount of characters in the string. Then the width of the word is being found with `computeMaxIntrinsicWidth`, and that is being set as the column width. This will cause text to be cutoff when the string found with the most amount of characters uses on average thiner characters than a shorter word.

This can be seen visually:

"IIIII"
Has six characters.

"WWWWW"
Has five characters but is much wider.

So we need to find a different way to get the largest character. We do not want to use `computeMaxIntrinsicWidth` more than we are now, as that's expensive. Ideally we would find a way to determine if a string is wider than another before any are used with `computeMaxIntrinsicWidth`. We could probably solve the clipping by adding a buffer space, but that would make all datepickers wider, even those which are not currently cropped.

**Comment by JCKodel** [comment]
I've noticed this problem and, at least in portuguese, it's caused by the letter M:

```
Fevereiro
Novembro
```

February has 1 letter more than November, but november has a M that is very wide, hence the bug:

<img width="385" alt="image" src="https://github.com/flutter/flutter/assets/379339/3b2c1d6f-56f8-4dab-869a-20778b93cd68">

> We do not want to use `computeMaxIntrinsicWidth` more than we are now, as that's expensive. 
>
I could suggest two solutions:

1) Compute all months width and cache it in a static map, using font name and font size as key (so the computation takes long at first run, but it runs smoothly afterwards).

2) Try to come up with a smarter `longestText` function that takes some premises such as: almost all of the time the font is variable width and most of the time it uses variable width fonts, measure each letter on a sans-serif font and assign a weight for each char (so I is 1, M is 5, for example), and then use those weights to determine the longest text. Not perfect, but at least, more accurate.

I have a very smart person here with me that came up with this code. His name is `G`e`p`e`t`o:

```dart
Map<String, double> characterWeights = {
  'i': 1.0, 'j': 1.0, 'l': 1.0,
  'f': 1.2, 'r': 1.2, 't': 1.2,
  'a': 1.5, 'c': 1.5, 'e': 1.5, 'o': 1.5, 's': 1.5, 'u': 1.5,
  'b': 1.7, 'd': 1.7, 'h': 1.7, 'k': 1.7, 'n': 1.7, 'p': 1.7,
  'g': 1.8, 'y': 1.8,
  'm': 2.2, 'w': 2.2,
  'Q': 2.2, 'W': 2.2, 'M': 2.2,
  ' ': 0.5, // space is narrower
};

double calculateStringWeight(String text) {
  var totalWeight = 0.0;
  
  for (int i = 0; i < text.length; i++) {
    final char = text[i];
    
    totalWeight += characterWeights[char] ?? 1.5;
  }
  
  print("${text} is ${totalWeight}");
  
  return totalWeight;
}

String findLongestText(List<String> texts) {
  var longestText = texts[0];
  var maxWeight = calculateStringWeight(longestText);

  for (var index = 1; index < texts.length; index++) {
    final text = texts[index];
    final weight = calculateStringWeight(text);
    
    if (weight > maxWeight) {
      maxWeight = weight;
      longestText = text;
    }
  }

  return longestText;
}

void main() {
  // Example usage:
  List<String> options = [
    "Janeiro",
    "Fevereiro",
    "Março",
    "Abril",
    "Maio",
    "Junho",
    "Julho",
    "Agosto",
    "Setembro",
    "Outubro",
    "Novembro",
    "Dezembro",
  ];

  final longestText = findLongestText(options);
  
  print('The longest text is: $longestText');
}
```

This would give this acceptable result (although maybe some more tests are required):

```
Janeiro is 9.9
Fevereiro is 12.399999999999999
Março is 7.9
Abril is 6.4
Maio is 6.2
Junho is 7.9
Julho is 7.2
Agosto is 9
Setembro is 12.299999999999999
Outubro is 10.1
Novembro is 12.599999999999998
Dezembro is 12.599999999999998

The longest text is: Novembro
```

And, of course, it would be a nice idea to cache those results, since the month names will not change and recalculations could be prevented with a simple `Map<_PickerColumnType, double>`.

**Comment by Mairramer** [comment]
@MitchellGoodwin Hi, I made an implementation that fixes this and also adds a way to calculate the widest word externally. I'm just having difficulty with the tests.

**Comment by Mairramer** [comment]
@MitchellGoodwin I thought of two possibilities:
1 - Calculate the width of each letter using painter and add it to the Map and then add the width of the largest character to painter.maxIntrinsicWidth:
However, it can be very costly depending on the case.
2 - Calculate the size of each word given a list, considering that we have 12 fixed words.

I need to invest a little more time to try new optimizations. I tested it with several languages ​​and it worked really well, but I still need to be more assertive with the tests.

**Comment by LongCatIsLooong** [comment]
>   calculate the widest word externally

Breaking up the text and measuring the width of each segment in isolation in general does not make sense. Since we're using the TextPainter API to paint the text anyways, `TextPainter.maxIntrinsicWidth` should give you the min width with which the text does not softwrap.

**Comment by LongCatIsLooong** [comment]
Hmm I looked at the implementation, the widget actually only uses the TextPainter for measuring purposes. In that case I think we should consider using `IntrinsicWidth`.

**Comment by Mairramer** [comment]
I made a simple solution for this. Take the 3 largest words and calculate their width. This solved the problem, as the scope decreased.

**Comment by LongCatIsLooong** [comment]
For tests, the test font we use is basically a monospace one (with a few exceptions), so the test issue is not a localization one, it's a font issue. If you can change the month text, you could use the codepoints with different advances in the [test font](https://github.com/flutter/flutter/blob/master/docs/contributing/testing/Flutter-Test-Fonts.md)

**Comment by MitchellGoodwin** [comment]
I'm not sure if there's a way to change the month text without going through localizations. [Localizations.override](https://github.com/flutter/flutter/blob/cb05b4e96d7a62866e3a693216b458290d463725/packages/flutter/lib/src/widgets/localizations.dart#L439) may be what's needed to get months that trigger this issue, combined with adjusting the test font.

**Comment by LongCatIsLooong** [comment]
> I'm not sure if there's a way to change the month text without going through localizations.

I think it's a case where we need to expose new test-only methods (https://api.flutter.dev/flutter/meta/visibleForTesting-constant.html) unfortunately. If you look at the test font glyph map only space characters have different x-advances.

**Comment by victorsanni** [comment]
I don't know if I should create a new issue for this, but I was running into something similar when testing #151494. 

With the locale `es`, in `_PickerColumnType.dayOfMonth`, some days of the week can't fit in one line in the column:

<img width="379" alt="date mode es locale incorrect with showDayofweek" src="https://github.com/flutter/flutter/assets/77553258/1cbf347b-6beb-4955-a326-278c8cee67cc">

**Comment by MitchellGoodwin** [comment]
It's odd that it overflows to two lines instead of getting cut off. But it is the same underlying issue of the picker wheels incorrectly deciding how wide they should be. #110319 appears to be the same issue as well. Maybe a new issue should be opened to represent all of these issues and we can close the others.

**Comment by victorsanni** [comment]
> I don't know if I should create a new issue for this, but I was running into something similar when testing #151494.
> 

This was fixed by #151128.
